src/instructlab/sdg/checkpointing.py

```diff
@@ -0,0 +1,93 @@
+# Standard
+import uuid
+
+# Third Party
+from datasets import Dataset, concatenate_datasets, load_dataset
+from datasets.data_files import EmptyDatasetError
+
+
+
+# First Party
+import logging
+
+logger = logging.getLogger(__name__)
+
+class Checkpointer:
+    def __init__(self, checkpoint_dir=None, save_freq=1):
+        self.checkpoint_dir = checkpoint_dir
+        # TODO: impliment this
+        self.save_freq = save_freq
+
+        self.cache = []
+
+    #def _save_intermediate_checkpoint(self, dataset, checkpoint_dir):
+    def save(self, dataset):
+        checkpoint_id = uuid.uuid4().hex
+        checkpoint_file = f"{self.checkpoint_dir}/data_checkpoint_{checkpoint_id}.jsonl"
+        logger.info(f"Saving checkpoint to {checkpoint_file}")
+        # Saves all the current records to new file in the checkpoint dir
+        dataset.to_json(checkpoint_file, orient="records", lines=True)
+
+    #def generate(self, dataset: Dataset, checkpoint_dir=None) -> Dataset:
+    def load(self, dataset: Dataset) -> Dataset:
+        # check if checkpoint_dir exists
+        seed_data = dataset
+        pre_generated_data = None
+        if self.checkpoint_dir is not None:
+            try:
+                # check if there are any existing checkpoints, loads all of the entries in the dir
+                # TODO(derekh): save should clear old entries so less is loaded
+                pre_generated_data = load_dataset(
+                    "json", data_dir=self.checkpoint_dir, split="train"
+                )
+                logger.info(
+                    f"Loading existing checkpoints from {self.checkpoint_dir}, with {pre_generated_data.num_rows} rows"
+                )
+                seed_data = self._get_missing_data(dataset, pre_generated_data)
+                logger.info(
+                    f"Found {seed_data.num_rows} missing rows in the dataset"
+                )
+
+            except EmptyDatasetError:
+                logger.info(
+                    f"No existing checkpoints found in {self.checkpoint_dir}, generating from scratch"
+                )
+        return seed_data, pre_generated_data
+
+
+    def _get_missing_data(self, seed_data, generated_data):
+
+        # Get the common columns between the two datasets
+        common_columns = list(
+            set(seed_data.column_names) & set(generated_data.column_names)
+        )
+
+        # Extract the relevant data based on common columns
+        seed_data_common = seed_data.select_columns(common_columns)
+        generated_data_common = generated_data.select_columns(common_columns)
+
+        # Convert to Pandas DataFrames for easier comparison
+        seed_df = seed_data_common.to_pandas()
+        generated_df = generated_data_common.to_pandas()
+
+        # Identify missing rows
+        missing_df = seed_df[
```

Comment:
Ok, this is identifying missing rows - but it looks like it will only return the common columns for those missing rows. What if the pipeline had deleted one of the original columns?

Should we instead be selecting from `seed_data`

the whole thing is quite confusing. Here's my attempt to make it more clear

```
def to_tuples(ds, columns):
    """Convert the dataset into a dataframe of tuples - one tuple per row - but only including the specified columns"""
   df = ds.select_columns(columns).to_pandas()
   return df.apply(tuple, axis=1)

# Get the common columns between the two datasets
common_columns = list(
    set(seed_data.column_names) & set(generated_data.column_names)
)

seed_tuples = to_tuples(seed_data, common_columns)
generated_tuples = to_tuples(generated_data, common_columns)

missing_rows = ~seed_tuples.isin(generated_tuples)

missing_df = seed_data.to_pandas()[missing_rows]

return utils.pandas.dataset_from_pandas_dataframe(missing_df)
```

Reply:
> Ok, this is identifying missing rows - but it looks like it will only return the common columns for those missing rows. What if the pipeline had deleted one of the original columns?

In that case we would have less columns in common,  which I guess would result in samples being marked as checkpointed in error.

> Should we instead be selecting from seed_data

yes, this would make sure a column that was deleted by a pipe would be in the returned dataset  but I think wouldn't help if samples were marked as checkpointed in error because of that missing column (I think that makes sense)


There is also a problem with rows that were filtered out by the pipeline, these never get checkpointed so will all be retried on subsequent runs

Reply:
Take a look at https://github.com/instructlab/sdg/pull/222/commits/ca11adf965879740172a2a9142c4553627f2dc14

I think that's a genuine issue?

Reply:
Yup, agreed, reproduced it locally with a rename block and your patch fixed it